File: pipeline/landfall_pipeline/generation/buckets.py
"""Bucket taxonomy, frozen per docs/decisions.md.

Citizenship resolves into independent rule buckets rather than being
treated as one 190-value variable (docs/prd.md, "Personalisation model").
The actual per-country -> bucket mapping is content knowledge that belongs
to the Week 1 pilot bucket, not hardcoded logic — this module only defines
the taxonomy shape and the realistic combinations to generate against.

Frozen 2026-09-24 (docs/decisions.md item 15) at 7 signatures — one more
than the PRD's rough "six realistic combinations" — derived directly from
real IRCC eTA/biometrics/medical-exam data researched for 21 citizenships
(the same table backing apps/web/src/lib/buckets.ts's client-side lookup;
keep the two in sync). #1 is the Week 1 pilot bucket, already published.
"""

from __future__ import annotations

from landfall_pipeline.store.schema import BucketSignature

REALISTIC_COMBINATIONS: list[BucketSignature] = [
    # 1. China, India, Nigeria, Vietnam, Bangladesh, Pakistan, Philippines,
    #    Indonesia. Already published (Week 1 pilot).
    BucketSignature(
        entry_document="visa_required",
        biometrics="required",
        medical_exam="required",
        funds_evidence="country_programme_variant",
        currency_corridor="restricted",
    ),
    # 2. United States.
    BucketSignature(
        entry_document="eta",
        biometrics="exempt",
        medical_exam="not_required",
        funds_evidence="standard",
        currency_corridor="major",
    ),
    # 3. Iran.
    BucketSignature(
        entry_document="visa_required",
        biometrics="required",
        medical_exam="not_required",
        funds_evidence="country_programme_variant",
        currency_corridor="restricted",
    ),
    # 4. South Korea, Hong Kong.
    BucketSignature(
        entry_document="eta",
        biometrics="required",
        medical_exam="required",
        funds_evidence="standard",
        currency_corridor="major",
    ),
    # 5. France, Japan, United Kingdom, Germany, Taiwan, Australia.
    BucketSignature(
        entry_document="eta",
        biometrics="required",
        medical_exam="not_required",
        funds_evidence="standard",
        currency_corridor="major",
    ),
    # 6. Mexico, Saudi Arabia.
    BucketSignature(
        entry_document="visa_required",
        biometrics="required",
        medical_exam="not_required",
        funds_evidence="standard",
        currency_corridor="major",
    ),
    # 7. Brazil.
    BucketSignature(
        entry_document="visa_required",
        biometrics="required",
        medical_exam="required",
        funds_evidence="standard",
        currency_corridor="major",
    ),
]


def known_signatures() -> set[str]:
    return {combo.signature() for combo in REALISTIC_COMBINATIONS}


def is_covered(signature: BucketSignature) -> bool:
    """False means: per docs/decisions.md, this is a taxonomy-coverage gap
    to route through the generation pipeline for expansion — not a
    permanent "closest match" fallback. The Next.js app's guide-level
    unmatched-signature state (docs/decisions.md item 14, apps/web/src/lib/
    guide.ts's nearestSignature()) handles the interim rendering."""
    return signature.signature() in known_signatures()

File: apps/web/src/lib/buckets.ts
// Citizenship -> bucket resolution (client-side, architecture.md §5).
//
// Real IRCC data researched 2026-09-24 (eTA-eligible list, biometrics
// exemptions, and the Designated Countries/Territories medical-exam
// table fetched directly from canada.ca) for a representative set of
// UBC's actual top source countries (China, India, USA, Iran per UBC's
// own enrolment report) plus other common ones. This is NOT exhaustive
// across all ~190 citizenships — an unlisted country falls back to
// DEFAULT_BUCKET, which is deliberately the same signature as the one
// real published pilot guide, so unlisted citizenships still resolve to
// real content rather than the "unmatched signature" fallback (SAA-22).
//
// funds_evidence and currency_corridor are coarser simplifications (see
// docs/prd.md's bucket table) without an equally authoritative single
// source — flagged here rather than presented as equally verified.

export type EntryDocument = "eta" | "visa_required";
export type Biometrics = "required" | "exempt";
export type MedicalExam = "required" | "not_required";
export type FundsEvidence = "standard" | "country_programme_variant";
export type CurrencyCorridor = "major" | "restricted";

export interface BucketSignature {
  entry_document: EntryDocument;
  biometrics: Biometrics;
  medical_exam: MedicalExam;
  funds_evidence: FundsEvidence;
  currency_corridor: CurrencyCorridor;
}

export function signatureString(b: BucketSignature): string {
  return [
    b.entry_document,
    b.biometrics,
    b.medical_exam,
    b.funds_evidence,
    b.currency_corridor,
  ].join("-");
}

// Same signature as pipeline/run_pilot.py's PILOT_BUCKET — the one
// bucket with a real published guide today.
export const DEFAULT_BUCKET: BucketSignature = {
  entry_document: "visa_required",
  biometrics: "required",
  medical_exam: "required",
  funds_evidence: "country_programme_variant",
  currency_corridor: "restricted",
};

export interface CitizenshipEntry {
  code: string;
  label: string;
  aliases: string[];
  bucket: BucketSignature;
}

export const CITIZENSHIPS: CitizenshipEntry[] = [
  { code: "CN", label: "China", aliases: ["PRC", "Mainland China"], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "IN", label: "India", aliases: ["Bharat"], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "US", label: "United States", aliases: ["USA", "US", "America"], bucket: { entry_document: "eta", biometrics: "exempt", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "IR", label: "Iran", aliases: ["Persia"], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "not_required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "KR", label: "South Korea", aliases: ["Korea", "Republic of Korea"], bucket: { entry_document: "eta", biometrics: "required", medical_exam: "required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "HK", label: "Hong Kong", aliases: ["Hong Kong SAR"], bucket: { entry_document: "eta", biometrics: "required", medical_exam: "required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "NG", label: "Nigeria", aliases: [], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "FR", label: "France", aliases: [], bucket: { entry_document: "eta", biometrics: "required", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "VN", label: "Vietnam", aliases: ["Viet Nam"], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "MX", label: "Mexico", aliases: [], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "BR", label: "Brazil", aliases: [], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "BD", label: "Bangladesh", aliases: [], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "JP", label: "Japan", aliases: [], bucket: { entry_document: "eta", biometrics: "required", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "GB", label: "United Kingdom", aliases: ["UK", "Britain", "Great Britain"], bucket: { entry_document: "eta", biometrics: "required", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "PK", label: "Pakistan", aliases: [], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "PH", label: "Philippines", aliases: [], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "DE", label: "Germany", aliases: [], bucket: { entry_document: "eta", biometrics: "required", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "ID", label: "Indonesia", aliases: [], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "required", funds_evidence: "country_programme_variant", currency_corridor: "restricted" } },
  { code: "SA", label: "Saudi Arabia", aliases: ["KSA"], bucket: { entry_document: "visa_required", biometrics: "required", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "TW", label: "Taiwan", aliases: [], bucket: { entry_document: "eta", biometrics: "required", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
  { code: "AU", label: "Australia", aliases: [], bucket: { entry_document: "eta", biometrics: "required", medical_exam: "not_required", funds_evidence: "standard", currency_corridor: "major" } },
];

export function searchCitizenships(query: string): CitizenshipEntry[] {
  const q = query.trim().toLowerCase();
  if (!q) return CITIZENSHIPS;
  return CITIZENSHIPS.filter(
    (c) =>
      c.label.toLowerCase().includes(q) ||
      c.code.toLowerCase() === q ||
      c.aliases.some((a) => a.toLowerCase().includes(q))
  );
}

export function resolveBucket(code: string): BucketSignature {
  const entry = CITIZENSHIPS.find((c) => c.code === code);
  return entry ? entry.bucket : DEFAULT_BUCKET;
}
